src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs

```diff
@@ -6,127 +6,103 @@
+                            _logger.LogError(exception, "Health check '{healthCheckName}' returned a result with a status of Unknown", healthCheck.Name);
+                            throw exception;
+                        }
+
+                        results[healthCheck.Name] = result;
+                    }
                 }
+
+                return new CompositeHealthCheckResult(results);
             }
         }
 
-        /// <summary>
-        /// Runs all the health checks in the application and returns the aggregated status.
-        /// </summary>
-        /// <param name="cancellationToken">A <see cref="CancellationToken"/> which can be used to cancel the health checks.</param>
-        /// <returns>
-        /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
-        /// yielding a <see cref="CompositeHealthCheckResult"/> containing the results.
-        /// </returns>
-        public Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default) =>
-            CheckHealthAsync(Checks.Values, cancellationToken);
-        
-        /// <summary>
-        /// Runs the provided health checks and returns the aggregated status
-        /// </summary>
-        /// <param name="checks">The <see cref="IHealthCheck"/> instances to be run.</param>
-        /// <param name="cancellationToken">A <see cref="CancellationToken"/> which can be used to cancel the health checks.</param>
-        /// <returns>
-        /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
-        /// yielding a <see cref="CompositeHealthCheckResult"/> containing the results.
-        /// </returns>
-        public async Task<CompositeHealthCheckResult> CheckHealthAsync(IEnumerable<IHealthCheck> checks, CancellationToken cancellationToken = default)
+        private static void EnsureNoDuplicates(IEnumerable<IHealthCheck> healthChecks)
         {
-            var results = new Dictionary<string, HealthCheckResult>(Checks.Count, StringComparer.OrdinalIgnoreCase);
-            foreach (var check in checks)
+            // Scan the list for duplicate names to provide a better error if there are duplicates.
+            var names = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            var duplicates = new List<string>();
+            foreach (var check in healthChecks)
             {
-                cancellationToken.ThrowIfCancellationRequested();
-                // If the health check does things like make Database queries using EF or backend HTTP calls,
-                // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
-                using (_logger.BeginScope(new HealthCheckLogScope(check.Name)))
+                if (!names.Add(check.Name))
```

Comment:
This would print the same name multiple times if the check is registered more than twice. Think a `GroupBy` would be terse and more correct

Reply:
sure good point.